File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def DFS(self , tree,path=[])-> list:
        if tree.left!=None:
            self.DFS(tree.left,path)
        if tree.val!=None:
            path.append(tree.val)
        if tree.right!=None:
            self.DFS(tree.right,path)
        return path
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root:
            return self.DFS(root,[])
        
        

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
class Solution {
private:
    vector<int> ans;
    void inorder(TreeNode* root){
        if(!root) return;
        
        TreeNode* curr = root, *pre;
        while(curr){
            if(!curr->left){
                ans.push_back(curr->val); curr = curr->right;
            }
            else{
                pre = curr->left;
                while(pre->right != NULL and pre->right != curr){
                    pre = pre->right;
                }
                if(pre->right == NULL){
                    pre->right = curr;
                    curr = curr->left;
                }
                else{
                    pre->right = NULL;
                    ans.push_back(curr->val);
                    curr = curr->right;
                }
            }
        }
        return;
    }
public:
    vector<int> inorderTraversal(TreeNode* root) {
        inorder(root);
        return ans;
    }
};